packages/core/src/common/uri-command-handler.ts

```diff
@@ -47,82 +46,128 @@ export interface MultiUriCommandHandler extends UriCommandHandler<URI[]> {
+        const [maybeURI, ...others] = args;
+        if (maybeURI instanceof URI) {
+            return [maybeURI, ...others];
+        }
+        const { selection } = this.selectionService;
+        const URIFromSelection = UriSelection.getUri(selection);
+        return [URIFromSelection, ...args];
+    }
+}
+
+export class MultiSelectUriCommandHandler extends AbstractUriAwareCommandHandler<URI[]> {
+    protected getArgsWithURI(...args: any[]): [URI[] | undefined, ...any[]] {
+        const [maybeURIs, ...others] = args;
+        if (Array.isArray(maybeURIs) && maybeURIs.every(uri => uri instanceof URI)) {
+            return [maybeURIs, ...others];
+        }
+        const { selection } = this.selectionService;
+        const URIsFromSelection = UriSelection.getUris(selection);
+        return [URIsFromSelection, ...args];
+    }
+}
+
+export class UriAwareCommandHandler<T extends MaybeArray<URI>> extends AbstractUriAwareCommandHandler<T> {
+    constructor(
+        protected readonly selectionService: SelectionService,
+        protected readonly handler: UriCommandHandler<T>,
+        protected readonly options?: UriAwareCommandHandler.Options
+    ) {
+        super(selectionService, handler);
+    }
+
+    protected getUri(...args: any[]): T | undefined {
+        const [maybeURIArray] = args;
+        const firstArgIsOK = this.isMulti()
+            ? Array.isArray(maybeURIArray) && maybeURIArray.every(uri => uri instanceof URI)
+            : maybeURIArray instanceof URI;
+
+        if (firstArgIsOK) {
+            return maybeURIArray;
+        }
+
+        const { selection } = this.selectionService;
+
+        const uriOrUris = this.isMulti()
+            ? UriSelection.getUris(selection)
+            : UriSelection.getUri(selection);
+
+        return uriOrUris as T;
+    }
+
+    protected getArgsWithURI(...args: any[]): [T | undefined, ...any[]] {
+        const uri = this.getUri(...args);
+        const [maybeURI, ...others] = args;
+        if (uri === maybeURI) {
+            return [maybeURI, ...others];
+        }
+        return [uri, ...args];
+    }
 
     protected isMulti(): boolean | undefined {
         return this.options && !!this.options.multi;
     }
+}
+
+export namespace UriAwareCommandHandler {
+
+    /**
+     * Further options for the URI aware command handler instantiation.
+     */
+    export interface Options {
+
+        /**
+         * `true` if the handler supports multiple selection. Otherwise, `false`. Defaults to `false`.
+         */
+        readonly multi?: boolean,
 
+    }
+
+    export function MultiSelect(selectionService: SelectionService, handler: UriCommandHandler<URI[]>): MultiSelectUriCommandHandler {
+        return new MultiSelectUriCommandHandler(selectionService, handler);
+    }
+
+    export function MonoSelect(selectionService: SelectionService, handler: UriCommandHandler<URI>): MonoSelectUriCommandHandler {
```

Comment:
Maybe add some jsdoc to explain what this kind of handler is supposed to do? (this is a public API)

Reply:
Good idea. I've added a JSDoc explaining what each handler should expect as the first parameter of its methods.